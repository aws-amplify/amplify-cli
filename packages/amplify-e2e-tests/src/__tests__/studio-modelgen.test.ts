import {
  initAndroidProjectWithProfile,
  addApiWithBlankSchema,
  addApiWithBlankSchemaAndConflictDetection,
  createNewProjectDir,
  deleteProject,
  deleteProjectDir,
  enableAdminUI,
  getLocalEnvInfo,
  getProjectMeta,
  amplifyPush,
  updateApiSchema,
  getDeploymentBucketObject,
  getProjectConfig,
} from '@aws-amplify/amplify-e2e-core';

describe('upload Studio CMS assets on push', () => {
  let projRoot: string;
  beforeEach(async () => {
    projRoot = await createNewProjectDir('studio-cms-upload');
  });

  afterEach(async () => {
    await deleteProject(projRoot);
    deleteProjectDir(projRoot);
  });
  it.each([
    [true, true],
    [false, true],
    [false, false],
  ])(
    'uploads expected CMS assets to shared location in S3 bucket %p %p',
    async (isStudioEnabled: boolean, apiWithConflictDetection: boolean) => {
      // eslint-disable-next-line spellcheck/spell-checker
      const name = `cms${isStudioEnabled}${apiWithConflictDetection}`;
      const defaultsSettings = {
        disableAmplifyAppCreation: false,
        name,
      };
      // init an android project to check that studio modelgen generates JS types even with other frontend config
      await initAndroidProjectWithProfile(projRoot, defaultsSettings);

      const originalProjectConfig = getProjectConfig(projRoot);

      const meta = getProjectMeta(projRoot);
      const appId = meta.providers?.awscloudformation?.AmplifyAppId;
      const region = meta.providers?.awscloudformation?.Region;

      const localEnvInfo = getLocalEnvInfo(projRoot);
      const { envName } = localEnvInfo;

      if (isStudioEnabled) {
        expect(appId).toBeDefined();
        // setup Amplify Studio backend
        await enableAdminUI(appId, envName, region);
      }

      if (apiWithConflictDetection) {
        await addApiWithBlankSchemaAndConflictDetection(projRoot, { transformerVersion: 2 });
      } else {
        await addApiWithBlankSchema(projRoot, { transformerVersion: 2 });
      }
      await updateApiSchema(projRoot, name, 'simple_model.graphql');
      await amplifyPush(projRoot);

      // expect CMS assets to be present in S3
      await expect(getDeploymentBucketObject(projRoot, `models/${name}/schema.graphql`)).resolves.toMatchInlineSnapshot(`
            "type Todo @model {
              id: ID!
              content: String
            }
            "
          `);
      await expect(getDeploymentBucketObject(projRoot, `models/${name}/schema.js`)).resolves.toMatchSnapshot();
      await expect(getDeploymentBucketObject(projRoot, `models/${name}/modelIntrospection.json`)).resolves.toMatchSnapshot();

      // expect project config to be unmodified
      expect(getProjectConfig(projRoot)).toEqual(originalProjectConfig);
    },
  );
});
