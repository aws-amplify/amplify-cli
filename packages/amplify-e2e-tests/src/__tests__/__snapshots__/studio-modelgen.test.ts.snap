// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload Studio CMS assets on push uploads expected CMS assets to shared location in S3 bucket false false 2`] = `
"export const schema = {
    "models": {
        "Todo": {
            "name": "Todo",
            "fields": {
                "id": {
                    "name": "id",
                    "isArray": false,
                    "type": "ID",
                    "isRequired": true,
                    "attributes": []
                },
                "content": {
                    "name": "content",
                    "isArray": false,
                    "type": "String",
                    "isRequired": false,
                    "attributes": []
                },
                "createdAt": {
                    "name": "createdAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                },
                "updatedAt": {
                    "name": "updatedAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                }
            },
            "syncable": true,
            "pluralName": "Todos",
            "attributes": [
                {
                    "type": "model",
                    "properties": {}
                }
            ]
        }
    },
    "enums": {},
    "nonModels": {},
    "codegenVersion": "3.4.4",
    "version": "6f85ca0ee370ac8e0674a44cc5a4f6a6"
};"
`;

exports[`upload Studio CMS assets on push uploads expected CMS assets to shared location in S3 bucket false false 3`] = `
"{
    "version": 1,
    "models": {
        "Todo": {
            "name": "Todo",
            "fields": {
                "id": {
                    "name": "id",
                    "isArray": false,
                    "type": "ID",
                    "isRequired": true,
                    "attributes": []
                },
                "content": {
                    "name": "content",
                    "isArray": false,
                    "type": "String",
                    "isRequired": false,
                    "attributes": []
                },
                "createdAt": {
                    "name": "createdAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                },
                "updatedAt": {
                    "name": "updatedAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                }
            },
            "syncable": true,
            "pluralName": "Todos",
            "attributes": [
                {
                    "type": "model",
                    "properties": {}
                }
            ],
            "primaryKeyInfo": {
                "isCustomPrimaryKey": false,
                "primaryKeyFieldName": "id",
                "sortKeyFieldNames": []
            }
        }
    },
    "enums": {},
    "nonModels": {}
}"
`;

exports[`upload Studio CMS assets on push uploads expected CMS assets to shared location in S3 bucket false true 2`] = `
"export const schema = {
    "models": {
        "Todo": {
            "name": "Todo",
            "fields": {
                "id": {
                    "name": "id",
                    "isArray": false,
                    "type": "ID",
                    "isRequired": true,
                    "attributes": []
                },
                "content": {
                    "name": "content",
                    "isArray": false,
                    "type": "String",
                    "isRequired": false,
                    "attributes": []
                },
                "createdAt": {
                    "name": "createdAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                },
                "updatedAt": {
                    "name": "updatedAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                }
            },
            "syncable": true,
            "pluralName": "Todos",
            "attributes": [
                {
                    "type": "model",
                    "properties": {}
                }
            ]
        }
    },
    "enums": {},
    "nonModels": {},
    "codegenVersion": "3.4.4",
    "version": "6f85ca0ee370ac8e0674a44cc5a4f6a6"
};"
`;

exports[`upload Studio CMS assets on push uploads expected CMS assets to shared location in S3 bucket false true 3`] = `
"{
    "version": 1,
    "models": {
        "Todo": {
            "name": "Todo",
            "fields": {
                "id": {
                    "name": "id",
                    "isArray": false,
                    "type": "ID",
                    "isRequired": true,
                    "attributes": []
                },
                "content": {
                    "name": "content",
                    "isArray": false,
                    "type": "String",
                    "isRequired": false,
                    "attributes": []
                },
                "createdAt": {
                    "name": "createdAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                },
                "updatedAt": {
                    "name": "updatedAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                }
            },
            "syncable": true,
            "pluralName": "Todos",
            "attributes": [
                {
                    "type": "model",
                    "properties": {}
                }
            ],
            "primaryKeyInfo": {
                "isCustomPrimaryKey": false,
                "primaryKeyFieldName": "id",
                "sortKeyFieldNames": []
            }
        }
    },
    "enums": {},
    "nonModels": {}
}"
`;

exports[`upload Studio CMS assets on push uploads expected CMS assets to shared location in S3 bucket true true 2`] = `
"export const schema = {
    "models": {
        "Todo": {
            "name": "Todo",
            "fields": {
                "id": {
                    "name": "id",
                    "isArray": false,
                    "type": "ID",
                    "isRequired": true,
                    "attributes": []
                },
                "content": {
                    "name": "content",
                    "isArray": false,
                    "type": "String",
                    "isRequired": false,
                    "attributes": []
                },
                "createdAt": {
                    "name": "createdAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                },
                "updatedAt": {
                    "name": "updatedAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                }
            },
            "syncable": true,
            "pluralName": "Todos",
            "attributes": [
                {
                    "type": "model",
                    "properties": {}
                }
            ]
        }
    },
    "enums": {},
    "nonModels": {},
    "codegenVersion": "3.4.4",
    "version": "6f85ca0ee370ac8e0674a44cc5a4f6a6"
};"
`;

exports[`upload Studio CMS assets on push uploads expected CMS assets to shared location in S3 bucket true true 3`] = `
"{
    "version": 1,
    "models": {
        "Todo": {
            "name": "Todo",
            "fields": {
                "id": {
                    "name": "id",
                    "isArray": false,
                    "type": "ID",
                    "isRequired": true,
                    "attributes": []
                },
                "content": {
                    "name": "content",
                    "isArray": false,
                    "type": "String",
                    "isRequired": false,
                    "attributes": []
                },
                "createdAt": {
                    "name": "createdAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                },
                "updatedAt": {
                    "name": "updatedAt",
                    "isArray": false,
                    "type": "AWSDateTime",
                    "isRequired": false,
                    "attributes": [],
                    "isReadOnly": true
                }
            },
            "syncable": true,
            "pluralName": "Todos",
            "attributes": [
                {
                    "type": "model",
                    "properties": {}
                }
            ],
            "primaryKeyInfo": {
                "isCustomPrimaryKey": false,
                "primaryKeyFieldName": "id",
                "sortKeyFieldNames": []
            }
        }
    },
    "enums": {},
    "nonModels": {}
}"
`;
